File: frontend/src/utils/shippingRequestUtils.ts
// Điều kiện để thao tác với Shipping Request của user
export const canEditUserShippingRequest = (value: string) => {
  return ['PENDING'].includes(value);
};

export const canCancelUserShippingRequest = (value: string) => {
  return ['PENDING'].includes(value);
};

export const canEmptyTrackingNumberShippingRequest = (value: string) => {
  return ['INQUIRY', 'COMPLAINT'].includes(value);
};

export const canEmptyRequestContentShippingRequest = (value: string) => {
  return ['DELIVERY_REMINDER', 'PICKUP_REMINDER'].includes(value);
};

// Điều kiện để thao tác với Shipping Request của manager
export const canProcessingManagerShippingRequest = (value: string) => {
  return ['PENDING', 'PROCESSING'].includes(value);
};

export const SHIPPING_REQUEST_MESSAGES: Record<string, string> = {
  INQUIRY: "Yêu cầu hỗ trợ - Không bắt buộc mã đơn hàng",
  COMPLAINT: "Yêu cầu khiếu nại - Không bắt buộc mã đơn hàng",
  DELIVERY_REMINDER: "Hối giao hàng - Nội dung không bắt buộc",
  CHANGE_ORDER_INFO: "Thay đổi thông tin đơn hàng - Bắt buộc nhập mã đơn hàng và nội dung",
  PICKUP_REMINDER: "Hối lấy hàng - Nội dung không bắt buộc",
};

export const getShippingRequestMessage = (type?: string): string | undefined => {
  if (!type) return undefined;
  return SHIPPING_REQUEST_MESSAGES[type];
};

// Manager: các trạng thái được phép chuyển tiếp tùy theo status hiện tại
export const getAllowedManagerStatuses = (currentStatus?: string): string[] => {
  if (!currentStatus) return [];

  switch (currentStatus) {
    case 'PENDING':
      return ['PROCESSING', 'RESOLVED', 'REJECTED'];

    case 'PROCESSING':
      return ['RESOLVED', 'REJECTED'];

    case 'RESOLVED':
    case 'REJECTED':
    case 'CANCELLED':
      return [];

    default:
      return [];
  }
};

export const SHIPPING_REQUEST_TYPES = ['COMPLAINT', 'PICKUP_REMINDER', 'DELIVERY_REMINDER', 'CHANGE_ORDER_INFO', 'INQUIRY'] as const;

export const SHIPPING_REQUEST_STATUS = ['PENDING', 'PROCESSING', 'RESOLVED', 'REJECTED', 'CANCELLED'] as const;

export const SHIPPING_REQUEST_FILTER_SORT = [
  'NEWEST',
  'OLDEST'
] as const;

// Public 
export const SHIPPING_REQUEST_TYPES_PUBLIC = ['COMPLAINT', 'INQUIRY'] as const;

export const translateShippingRequestType = (value: string): string => {
  switch (value) {
    case 'COMPLAINT': return 'Khiếu nại';
    case 'PICKUP_REMINDER': return 'Hối lấy hàng';
    case 'DELIVERY_REMINDER': return 'Hối giao hàng';
    case 'CHANGE_ORDER_INFO': return 'Thay đổi thông tin ĐH';
    case 'INQUIRY': return 'Yêu cầu hỗ trợ';
    default: return value;
  }
};

export const translateShippingRequestStatus = (value: string): string => {
  switch (value) {
    case 'PENDING': return 'Chờ xử lý';
    case 'PROCESSING': return 'Đang xử lý';
    case 'RESOLVED': return 'Đã xử lý';
    case 'REJECTED': return 'Từ chối';
    case 'CANCELLED': return 'Đã hủy';
    default: return value;
  }
};

export const translateShippingRequestFilterSort = (value: string): string => {
  switch (value) {
    case 'NEWEST': return 'Mới nhất';
    case 'OLDEST': return 'Cũ nhất';
    default: return value;
  }
};

File: backend/src/main/java/com/logistics/utils/ShippingRequestUtils.java
package com.logistics.utils;

import com.logistics.entity.Office;
import com.logistics.enums.OrderStatus;
import com.logistics.enums.ShippingRequestStatus;
import com.logistics.enums.ShippingRequestType;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.EnumSet;
import java.util.Map;
import java.util.Set;

public class ShippingRequestUtils {

    private static final Map<ShippingRequestStatus, Set<ShippingRequestStatus>> MANAGER_ALLOWED_STATUS_TRANSITIONS = Map.of(
        ShippingRequestStatus.PENDING, Set.of(
            ShippingRequestStatus.PROCESSING,
            ShippingRequestStatus.RESOLVED, 
            ShippingRequestStatus.REJECTED
        ),
        ShippingRequestStatus.PROCESSING, Set.of(
            ShippingRequestStatus.RESOLVED,
            ShippingRequestStatus.REJECTED
        )
    );

    public static boolean canManagerChangeStatus(ShippingRequestStatus currentStatus, ShippingRequestStatus targetStatus) {
        if (currentStatus == null || targetStatus == null) return false;
        return MANAGER_ALLOWED_STATUS_TRANSITIONS.getOrDefault(currentStatus, Set.of()).contains(targetStatus);
    }

    // Những yêu cầu mà contentRequest được null
    private static final Set<ShippingRequestType> REQUEST_TYPE_ALLOWED_EMPTY_CONTENT_REQUEST = Set.of(
            ShippingRequestType.DELIVERY_REMINDER,
            ShippingRequestType.PICKUP_REMINDER);

    public static boolean canUserEmptyContentRequest(ShippingRequestType value) {
        return REQUEST_TYPE_ALLOWED_EMPTY_CONTENT_REQUEST.contains(value);
    }

    // Những yêu cầu mà trackingNumber được null
    private static final Set<ShippingRequestType> REQUEST_TYPE_ALLOWED_EMPTY_TRACKING_NUMBER = Set.of(
            ShippingRequestType.COMPLAINT,
            ShippingRequestType.INQUIRY);

    public static boolean canUserEmptyTrackingNumber(ShippingRequestType value) {
        return REQUEST_TYPE_ALLOWED_EMPTY_TRACKING_NUMBER.contains(value);
    }

    // Những yêu cầu có thể được thực hiện tùy vào trạng thái đơn hàng
    public static boolean isValidRequestForOrder(ShippingRequestType requestType, OrderStatus orderStatus) {
        switch (requestType) {
            case CHANGE_ORDER_INFO:
                return !EnumSet.of(
                        OrderStatus.DRAFT,
                        OrderStatus.DELIVERED,
                        OrderStatus.CANCELLED,
                        OrderStatus.RETURNED).contains(orderStatus);

            case DELIVERY_REMINDER:
                return EnumSet.of(
                        OrderStatus.AT_ORIGIN_OFFICE,
                        OrderStatus.IN_TRANSIT,
                        OrderStatus.AT_DEST_OFFICE,
                        OrderStatus.DELIVERING,
                        OrderStatus.RETURNING).contains(orderStatus);

            case PICKUP_REMINDER:
                return EnumSet.of(
                        OrderStatus.READY_FOR_PICKUP,
                        OrderStatus.CONFIRMED,
                        OrderStatus.PICKING_UP,
                        OrderStatus.PICKUP_RETRY,
                        OrderStatus.URGENT_PICKUP).contains(orderStatus);

            case COMPLAINT:
                return !EnumSet.of(
                        OrderStatus.DRAFT,
                        OrderStatus.PENDING).contains(orderStatus);

            default:
                return false;
        }
    }

    // Những yêu cầu có thể được thực hiện tùy vào trạng thái đơn hàng
    public static Office getValidOfficeForRequest(
            ShippingRequestType requestType,
            OrderStatus orderStatus,
            Office fromOffice,
            Office toOffice) {

        switch (requestType) {
            case CHANGE_ORDER_INFO:
                if (EnumSet.of(
                        OrderStatus.CONFIRMED,
                        OrderStatus.READY_FOR_PICKUP,
                        OrderStatus.PICKING_UP,
                        OrderStatus.PICKED_UP,
                        OrderStatus.AT_ORIGIN_OFFICE).contains(orderStatus)) {
                    return fromOffice;
                }
                if (EnumSet.of(
                        OrderStatus.IN_TRANSIT,
                        OrderStatus.AT_DEST_OFFICE,
                        OrderStatus.DELIVERING,
                        OrderStatus.RETURNING).contains(orderStatus)) {
                    return toOffice;
                }
                break;

            case DELIVERY_REMINDER:
                if (EnumSet.of(
                        OrderStatus.AT_ORIGIN_OFFICE,
                        OrderStatus.IN_TRANSIT).contains(orderStatus)) {
                    return fromOffice;
                }
                if (EnumSet.of(
                        OrderStatus.AT_DEST_OFFICE,
                        OrderStatus.DELIVERING,
                        OrderStatus.RETURNING).contains(orderStatus)) {
                    return toOffice;
                }
                break;

            case PICKUP_REMINDER:
                if (EnumSet.of(
                        OrderStatus.READY_FOR_PICKUP,
                        OrderStatus.PICKING_UP).contains(orderStatus)) {
                    return fromOffice;
                }
                break;
            default:
                return null;
        }

        return null;
    }

    public static String generateRequestCode(Integer id) {
        if (id == null) {
            throw new IllegalArgumentException("Id không được null khi tạo code");
        }
        String date = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyyMMdd"));
        return "SR" + date + id;
    }

    public static final Set<ShippingRequestStatus> ACTIVE_STATUSES = Set.of(ShippingRequestStatus.PENDING,
            ShippingRequestStatus.PROCESSING);

    // Những yêu cầu mà được hủy
    private static final Set<ShippingRequestStatus> REQUEST_TYPE_ALLOWED_CANCEL = Set.of(
            ShippingRequestStatus.PENDING);

    public static boolean canUserCancel(ShippingRequestStatus value) {
        return REQUEST_TYPE_ALLOWED_CANCEL.contains(value);
    }

    // Những yêu cầu mà guest có thể tạo
    private static final Set<ShippingRequestType> REQUEST_TYPE_ALLOWED_GUEST_CREATE = Set.of(
            ShippingRequestType.COMPLAINT,
            ShippingRequestType.INQUIRY);

    public static boolean canGuestCreateShippingRequest(ShippingRequestType value) {
        return REQUEST_TYPE_ALLOWED_GUEST_CREATE.contains(value);
    }

    public static String translateShippingRequestType(ShippingRequestType value) {
        if (value == null) {
            return "";
        }

        return switch (value) {
            case COMPLAINT -> "Khiếu nại";
            case PICKUP_REMINDER -> "Hối lấy hàng";
            case DELIVERY_REMINDER -> "Hối giao hàng";
            case CHANGE_ORDER_INFO -> "Thay đổi thông tin ĐH";
            case INQUIRY -> "Yêu cầu hỗ trợ";
            default -> value.name();
        };
    }

    public static String translateShippingRequestStatus(ShippingRequestStatus value) {
        if (value == null) {
            return "";
        }

        return switch (value) {
            case PENDING -> "Chờ xử lý";
            case PROCESSING -> "Đang xử lý";
            case RESOLVED -> "Đã xử lý";
            case REJECTED -> "Từ chối";
            case CANCELLED -> "Đã hủy";
            default -> value.name();
        };
    }
}